### PR TITLE
add survival analysis

### PR DESCRIPTION
**Pull request by kaungmyatwaiyan**
## Pull Request Contents

| |
|-|
✨ Feature

### Perform Survival Analysis
- Accepts Vocab ID and concept name as inputs
- Gets associated concept ID 
- Calculates the duration of the condition 
- Performs survival analysis
- Generates KM plot with median survival line and saves as png

## Related Issues or other material

- [ ] This PR relates to one or more issues, detailed below

- Related Issue #
- Closes #

## ✅ Added/updated tests?

- [ ] This PR contains relevant tests
  - Or doesn't need to per the below explanation

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

- [x] I've completed all **actions** and **tasks** detailed in the PR Template

### CODE REVIEW

File: app/beacon-omop-worker/beacon_omop_worker/query_solvers.py [hunk 1]
```diff
@@ -307,3 +309,167 @@ def solve_individuals(db_manager: SyncDBManager, query_terms: str):
...
+                concept ON condition_occurrence.condition_concept_id = concept.concept_id
+            JOIN
+                person ON condition_occurrence.person_id = person.person_id
+            JOIN
+                location ON person.location_id = location.location_id
+            LEFT JOIN
+                death ON condition_occurrence.person_id = death.person_id
+            WHERE
+                condition_occurrence.condition_concept_id = :concept_id
+            """
```
Comment (dcl10) [inline app/beacon-omop-worker/beacon_omop_worker/query_solvers.py:383]: We've found in previous query types that it is best to load the tables into memory and then perform the joins. The database is simply too slow. What you should do is load the individual tables into pandas, limiting the data from those tables with a `WHERE` clause if possible. Then join the tables using pandas.

File: app/beacon-omop-worker/beacon_omop_worker/query_solvers.py [hunk 2]
```diff
@@ -307,3 +309,167 @@ def solve_individuals(db_manager: SyncDBManager, query_terms: str):
...
+            plt.legend()
+
+            # Save the plot to the specified file
+            plt.savefig(file_path, format='png')  # You can change 'png' to 'pdf' if you prefer
+            plt.close()  # Close the plot to prevent display issues in non-GUI environments
+
+            logger.info(f"Kaplan-Meier plot saved to {file_path} for concept_id {concept_id}.")
+        except Exception as e:
+            logger.error(f"Error while plotting and saving Kaplan-Meier curve: {e}")
+            raise
```
Comment (dcl10) [inline app/beacon-omop-worker/beacon_omop_worker/query_solvers.py:447]:
```suggestion
            raise e
```
Raise the exception that was thrown

File: app/beacon-omop-worker/beacon_omop_worker/query_solvers.py [hunk 3]
```diff
@@ -307,3 +309,167 @@ def solve_individuals(db_manager: SyncDBManager, query_terms: str):
...
+        concept_id = solver._get_concept_id(vocab, code)
+
+        # Fetch condition data for the concept_id
+        condition_data = solver.fetch_condition_data(concept_id)
+
+        # Save Kaplan-Meier curve to file
+        solver.plot_kaplan_meier_curve(condition_data, concept_id, output_file)
+    except Exception as e:
+        logger.error(f"Error generating Kaplan-Meier plot for {snomed_code}: {e}")
+        raise
```
Comment (dcl10) [inline app/beacon-omop-worker/beacon_omop_worker/query_solvers.py:475]:
```suggestion
        raise e
```

**Review by dcl10 — CHANGES_REQUESTED** [review]

File: app/beacon-omop-worker/beacon_omop_worker/query_solvers.py (hunk 1 above)
Comment (kaungmyatwaiyan) [reply app/beacon-omop-worker/beacon_omop_worker/query_solvers.py]: Modified

File: app/beacon-omop-worker/beacon_omop_worker/query_solvers.py [hunk 4]
```diff
@@ -307,3 +309,203 @@ def solve_individuals(db_manager: SyncDBManager, query_terms: str):
...
+            concept_id (str): The concept_id for the condition.
+        
+        Returns:
+            pd.DataFrame: DataFrame containing the condition data.
+        """
+        logger = logging.getLogger(config.LOGGER_NAME)
+        
+        try:
+            # Load individual tables into pandas DataFrames with filtering
+            condition_occurrence_df = pd.read_sql(
```
Comment (dcl10) [inline app/beacon-omop-worker/beacon_omop_worker/query_solvers.py:359]: We have a table defined in `entities.py` for condition occurrences. You can actually create this query programmatically, similar to how it's done in `_get_concept_id`. Also you can check out the docs for sqlalchemy2 for extra help.

File: app/beacon-omop-worker/beacon_omop_worker/query_solvers.py [hunk 5]
```diff
@@ -307,3 +309,203 @@ def solve_individuals(db_manager: SyncDBManager, query_terms: str):
...
+                text("SELECT * FROM condition_occurrence WHERE condition_concept_id = :concept_id"),
+                con=self.db_manager.engine.connect(),
+                params={"concept_id": concept_id}
+            )
+            
+            concept_df = pd.read_sql(
```
Comment (dcl10) [inline app/beacon-omop-worker/beacon_omop_worker/query_solvers.py:365]: same here

File: app/beacon-omop-worker/beacon_omop_worker/query_solvers.py [hunk 6]
```diff
@@ -307,3 +309,203 @@ def solve_individuals(db_manager: SyncDBManager, query_terms: str):
...
+                text("SELECT concept_id, concept_name FROM concept"),
+                con=self.db_manager.engine.connect()
+            )
+            
+            person_df = pd.read_sql(
```
Comment (dcl10) [inline app/beacon-omop-worker/beacon_omop_worker/query_solvers.py:370]: same here

File: app/beacon-omop-worker/beacon_omop_worker/query_solvers.py [hunk 7]
```diff
@@ -307,3 +309,203 @@ def solve_individuals(db_manager: SyncDBManager, query_terms: str):
...
+                text("SELECT person_id, birth_datetime, gender_source_value, race_source_value, location_id FROM person"),
+                con=self.db_manager.engine.connect()
+            )
+            
+            location_df = pd.read_sql(
```
Comment (dcl10) [inline app/beacon-omop-worker/beacon_omop_worker/query_solvers.py:375]: You'll have to make this table. But do it in another PR and then change this bit here. This  bit is okay for now.

File: app/beacon-omop-worker/beacon_omop_worker/query_solvers.py [hunk 8]
```diff
@@ -307,3 +309,203 @@ def solve_individuals(db_manager: SyncDBManager, query_terms: str):
...
+                text("SELECT location_id, city, county FROM location"),
+                con=self.db_manager.engine.connect()
+            )
+            
+            death_df = pd.read_sql(
```
Comment (dcl10) [inline app/beacon-omop-worker/beacon_omop_worker/query_solvers.py:380]: Same for this table.

**Review by dcl10 — CHANGES_REQUESTED** [review]

**Review by dcl10 — APPROVED** [review]

**Review by vpnu — APPROVED** [review]
